pkg/cloudprovider/vsphere/loadbalancer/lbprovider.go

```diff
@@ -0,0 +1,182 @@
+func (p *lbProvider) GetLoadBalancerName(_ context.Context, clusterName string, service *corev1.Service) string {
+	return clusterName + ":" + service.Namespace + ":" + service.Name
+}
+
+// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
+// Implementations must treat the *corev1.Service and *corev1.Node
+// parameters as read-only and not modify them.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) EnsureLoadBalancer(_ context.Context, clusterName string, service *corev1.Service, nodes []*corev1.Node) (*corev1.LoadBalancerStatus, error) {
+	key := namespacedNameFromService(service).String()
+	p.keyLock.Lock(key)
+	defer p.keyLock.Unlock(key)
+
+	class, err := p.classFromService(service)
+	if err != nil {
+		return nil, err
+	}
+
+	state := newState(p.lbService, clusterName, service, nodes)
+	err = state.Process(class)
+	status, err2 := state.Finish()
+	if err != nil {
+		return status, err
+	}
+	return status, err2
+}
+
+func (p *lbProvider) classFromService(service *corev1.Service) (*loadBalancerClass, error) {
+	annos := service.GetAnnotations()
+	if annos == nil {
+		annos = map[string]string{}
+	}
+	name, ok := annos[LoadBalancerClassAnnotation]
+	name = strings.TrimSpace(name)
+	if !ok || name == "" {
+		name = config.DefaultLoadBalancerClass
+	}
+
+	class := p.classes.GetClass(name)
+	if class == nil {
+		return nil, fmt.Errorf("invalid load balancer class %s", name)
+	}
+	return class, nil
+}
+
+// UpdateLoadBalancer updates hosts under the specified load balancer.
+// Implementations must treat the *corev1.Service and *corev1.Node
+// parameters as read-only and not modify them.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) UpdateLoadBalancer(_ context.Context, clusterName string, service *corev1.Service, nodes []*corev1.Node) error {
+	key := namespacedNameFromService(service).String()
+	p.keyLock.Lock(key)
+	defer p.keyLock.Unlock(key)
+
+	state := newState(p.lbService, clusterName, service, nodes)
+
+	return state.UpdatePoolMembers()
+}
+
+// EnsureLoadBalancerDeleted deletes the specified load balancer if it
+// exists, returning nil if the load balancer specified either didn't exist or
+// was successfully deleted.
+// This construction is useful because many cloud providers' load balancers
+// have multiple underlying components, meaning a Get could say that the LB
+// doesn't exist even if some part of it is still laying around.
+// Implementations must treat the *corev1.Service parameter as read-only and not modify it.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (p *lbProvider) EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *corev1.Service) error {
+	emptyService := *service
+	emptyService.Spec.Ports = nil
```

Comment:
You shouldn't touch fields that come in from the cache. If you need to nil ports, do a `.DeepCopy()` on the service object.

Also wondering why we need to nil the ports field here?

Reply:
Good catch, thanks! This was an ugly bug.
The called EnsureLoadBalancer can create and delete virtual servers. It acts on the delta to the existing state. Therefore removing all ports translates into removing all virtual servers (there is a virtual server for each service port). Additionally, in the managed mode the load balancer service is deleted after the last virtual server has been deleted.